lib/token-store.js

```diff
@@ -8,196 +8,1110 @@
+     * @param {number} count - The count of tokens this cursor iterates.
+     */
+    constructor(cursor, count) {
+        super(cursor);
+        this.count = count;
+    }
+
+    /** @inheritdoc */
+    moveNext() {
+        if (this.count > 0) {
+            this.count -= 1;
+            return super.moveNext();
+        }
+        return false;
+    }
+
+    /** @inheritdoc */
+    movePrev() {
+        if (this.count > 0) {
+            this.count -= 1;
+            return super.movePrev();
+        }
+        return false;
+    }
+}
+
+/**
+ * Creates the map from locations to indices in `tokens`.
+ *
+ * The first/last location of tokens is mapped to the index of the token.
+ * The first/last location of comments is mapped to the index of the previous token of each comment.
+ *
+ * @param {Token[]} tokens - The array of tokens.
+ * @param {Comment[]} comments - The array of comments.
+ * @returns {Object} The map from locations to indices in `tokens`.
+ * @private
+ */
+function createIndexMap(tokens, comments) {
+    const map = Object.create(null);
+    let it = 0;
+    let ic = 0;
+    let nextStart = 0;
+    let range = null;
+
+    while (it < tokens.length || ic < comments.length) {
+        nextStart = (ic < comments.length) ? comments[ic].range[0] : Number.MAX_SAFE_INTEGER;
+        while (it < tokens.length && (range = tokens[it].range)[0] < nextStart) {
+            map[range[0]] = it;
+            map[range[1] - 1] = it;
+            it += 1;
+        }
+
+        nextStart = (it < tokens.length) ? tokens[it].range[0] : Number.MAX_SAFE_INTEGER;
+        while (ic < comments.length && (range = comments[ic].range)[0] < nextStart) {
+            map[range[0]] = it;
+            map[range[1] - 1] = it;
+            ic += 1;
+        }
+    }
+
+    return map;
+}
+
+/**
+ * Creates the cursor iterates tokens with normalized options.
+ *
+ * @param {Token[]} tokens - The array of tokens.
+ * @param {Comment[]} comments - The array of comments.
+ * @param {Object} imap - The map from locations to indices in `this.tokens`.
+ * @param {number} start - The start location of the iteration range.
+ * @param {number} end - The end location of the iteration range.
+ * @param {boolean} includeComments - The flag to iterate comments as well.
+ * @param {Function|null} filter - The predicate function to choose tokens.
+ * @param {number} skip - The count of tokens the cursor skips.
+ * @param {number} count - The maximum count of tokens the cursor iterates. Zero is no iteration for backward compatibility.
+ * @returns {Cursor} The created cursor.
+ * @private
+ */
+function createCursor(tokens, comments, imap, start, end, includeComments, filter, skip, count) {
+    let cursor = new (includeComments ? TokenCommentCursor : TokenCursor)(tokens, comments, imap, start, end);
+
+    if (filter) {
+        cursor = new FilterCursor(cursor, filter);
+    }
+    if (skip >= 1) {
+        cursor = new SkipCursor(cursor, skip);
```

Comment:
I guess it would be nice to be able to go either way on this. But I'm not sure I can provide any use case to back up my preference.